File: bot/config/secret_manager.py
import os
import json
import logging
from google.cloud import secretmanager
from typing import Dict, Optional

logger = logging.getLogger(__name__)

class SecretManager:
    """
    Google Cloud Secret Manager client for retrieving secrets
    """
    
    def __init__(self):
        self.client = secretmanager.SecretManagerServiceClient()
        self.project_id = os.getenv('GOOGLE_CLOUD_PROJECT_ID')
        self.cache = {}
        self.cache_expiry = 5 * 60  # 5 minutes in seconds
        
    def get_secret(self, secret_name: str, version: str = 'latest') -> Optional[str]:
        """
        Get a secret from Google Cloud Secret Manager or environment variable
        
        Args:
            secret_name: Name of the secret
            version: Version of the secret (default: 'latest')
            
        Returns:
            The secret value as string
        """
        # First check environment variables (for simple deployment or local dev)
        env_value = os.getenv(secret_name.upper().replace('-', '_'))
        if env_value:
            logger.info(f"Using environment variable for {secret_name}")
            return env_value
        
        # Check cache first
        cache_key = f"{secret_name}:{version}"
        if cache_key in self.cache:
            cached_data = self.cache[cache_key]
            if (cached_data['timestamp'] + self.cache_expiry) > self._get_current_time():
                return cached_data['value']
        
        try:
            # Cloud Run environment - get from Secret Manager
            name = f"projects/{self.project_id}/secrets/{secret_name}/versions/{version}"
            response = self.client.access_secret_version(request={"name": name})
            secret = response.payload.data.decode("UTF-8")
            
            # Cache the secret
            self.cache[cache_key] = {
                'value': secret,
                'timestamp': self._get_current_time()
            }
            
            return secret
        except Exception as error:
            logger.error(f"Failed to retrieve secret {secret_name}: {error}")
            
            # Fallback to environment variable
            env_value = os.getenv(secret_name.upper())
            if env_value:
                return env_value
            
            raise Exception(f"Secret {secret_name} not found in Secret Manager or environment variables")
    
    def get_secrets(self, secret_names: list) -> Dict[str, Optional[str]]:
        """
        Get multiple secrets at once
        
        Args:
            secret_names: List of secret names
            
        Returns:
            Dictionary with secret names as keys and values as secrets
        """
        secrets = {}
        for name in secret_names:
            try:
                secrets[name] = self.get_secret(name)
            except Exception as error:
                logger.error(f"Failed to get secret {name}: {error}")
                secrets[name] = None
        
        return secrets
    
    def get_service_account_credentials(self) -> dict:
        """
        Get Google service account credentials
        
        Returns:
            Service account credentials as dictionary
        """
        try:
            # Check for base64-encoded credentials in environment variable first
            if os.getenv('GOOGLE_SERVICE_ACCOUNT_CREDENTIALS'):
                try:
                    import base64
                    decoded = base64.b64decode(os.getenv('GOOGLE_SERVICE_ACCOUNT_CREDENTIALS')).decode()
                    credentials = json.loads(decoded)
                    logger.info('Using base64-encoded service account credentials from environment')
                    return credentials
                except Exception as decode_error:
                    # If base64 decode fails, try as direct JSON
                    try:
                        credentials = json.loads(os.getenv('GOOGLE_SERVICE_ACCOUNT_CREDENTIALS'))
                        logger.info('Using direct JSON service account credentials from environment')
                        return credentials
                    except Exception as json_error:
                        logger.error(f"Failed to parse service account credentials from environment: {json_error}")

            # Try Secret Manager
            credentials_json = self.get_secret('google-service-account-credentials')
            return json.loads(credentials_json)
        except Exception as error:
            logger.error(f"Failed to get service account credentials: {error}")
            
            # For local development, try the GOOGLE_APPLICATION_CREDENTIALS file
            if os.getenv('GOOGLE_APPLICATION_CREDENTIALS') and os.getenv('GOOGLE_APPLICATION_CREDENTIALS').endswith('.json'):
                try:
                    with open(os.getenv('GOOGLE_APPLICATION_CREDENTIALS'), 'r') as f:
                        return json.load(f)
                except Exception as file_error:
                    logger.error(f"Failed to read local credentials file: {file_error}")
            
            raise Exception('Service account credentials not found')
    
    def clear_cache(self):
        """Clear the cache (useful for testing or forced refresh)"""
        self.cache.clear()
    
    def _get_current_time(self) -> int:
        """Get current timestamp"""
        import time
        return int(time.time())

# Create a singleton instance
secret_manager = SecretManager()

File: server/config/secretManager.js
const { SecretManagerServiceClient } = require('@google-cloud/secret-manager');

class SecretManager {
    constructor() {
        this.client = new SecretManagerServiceClient();
        this.projectId = process.env.GOOGLE_CLOUD_PROJECT || process.env.GOOGLE_CLOUD_PROJECT_ID;
        if (!this.projectId) {
            throw new Error('GOOGLE_CLOUD_PROJECT environment variable is not set. Unable to determine project ID for Secret Manager.');
        }
        console.log(`SecretManager initialized with projectId: ${this.projectId}`);
        this.cache = new Map();
        this.cacheExpiry = 5 * 60 * 1000; // 5 minutes
    }

    /**
     * Get a secret from Google Cloud Secret Manager or environment variable
     * @param {string} secretName - Name of the secret
     * @param {string} version - Version of the secret (default: 'latest')
     * @returns {Promise<string>} The secret value
     */
    async getSecret(secretName, version = 'latest') {
        // First check environment variables (for simple deployment or local dev)
        const envValue = process.env[secretName.toUpperCase()];
        if (envValue) {
            console.log(`Using environment variable for ${secretName}`);
            return envValue;
        }

        // Check cache first
        const cacheKey = `${secretName}:${version}`;
        const cached = this.cache.get(cacheKey);
        if (cached && (Date.now() - cached.timestamp) < this.cacheExpiry) {
            return cached.value;
        }

        try {
            // Cloud Run environment - get from Secret Manager
            const name = `projects/${this.projectId}/secrets/${secretName}/versions/${version}`;
            const [version_response] = await this.client.accessSecretVersion({ name });
            const secret = version_response.payload.data.toString();
            
            // Cache the secret
            this.cache.set(cacheKey, {
                value: secret,
                timestamp: Date.now()
            });
            
            return secret;
        } catch (error) {
            console.error(`Failed to retrieve secret ${secretName} for project ${this.projectId}:`, error.message);
            
            // Fallback to environment variable
            const envValue = process.env[secretName.toUpperCase()];
            if (envValue) {
                return envValue;
            }
            
            throw new Error(`Secret ${secretName} not found in Secret Manager or environment variables`);
        }
    }

    /**
     * Get multiple secrets at once
     * @param {string[]} secretNames - Array of secret names
     * @returns {Promise<Object>} Object with secret names as keys and values as secrets
     */
    async getSecrets(secretNames) {
        const secrets = {};
        const promises = secretNames.map(async (name) => {
            try {
                secrets[name] = await this.getSecret(name);
            } catch (error) {
                console.error(`Failed to get secret ${name}:`, error.message);
                secrets[name] = null;
            }
        });
        
        await Promise.all(promises);
        return secrets;
    }

    /**
     * Get Google service account credentials
     * @returns {Promise<Object>} Service account credentials object
     */
    async getServiceAccountCredentials() {
        try {
            // Check for base64-encoded credentials in environment variable first
            if (process.env.GOOGLE_SERVICE_ACCOUNT_CREDENTIALS) {
                try {
                    const decoded = Buffer.from(process.env.GOOGLE_SERVICE_ACCOUNT_CREDENTIALS, 'base64').toString();
                    const credentials = JSON.parse(decoded);
                    console.log('Using base64-encoded service account credentials from environment');
                    return credentials;
                } catch (decodeError) {
                    // If base64 decode fails, try as direct JSON
                    try {
                        const credentials = JSON.parse(process.env.GOOGLE_SERVICE_ACCOUNT_CREDENTIALS);
                        console.log('Using direct JSON service account credentials from environment');
                        return credentials;
                    } catch (jsonError) {
                        console.error('Failed to parse service account credentials from environment:', jsonError.message);
                    }
                }
            }

            // Try Secret Manager
            const credentialsJson = await this.getSecret('google-service-account-credentials');
            return JSON.parse(credentialsJson);
        } catch (error) {
            console.error('Failed to get service account credentials:', error.message);
            
            // For local development, try the GOOGLE_APPLICATION_CREDENTIALS file
            if (process.env.GOOGLE_APPLICATION_CREDENTIALS && process.env.GOOGLE_APPLICATION_CREDENTIALS.endsWith('.json')) {
                const fs = require('fs');
                try {
                    const credentialsFile = fs.readFileSync(process.env.GOOGLE_APPLICATION_CREDENTIALS, 'utf8');
                    return JSON.parse(credentialsFile);
                } catch (fileError) {
                    console.error('Failed to read local credentials file:', fileError.message);
                }
            }
            
            throw new Error('Service account credentials not found');
        }
    }

    /**
     * Clear the cache (useful for testing or forced refresh)
     */
    clearCache() {
        this.cache.clear();
    }
}

// Create a singleton instance
const secretManager = new SecretManager();

module.exports = secretManager;